securedrop/tests/functional/test_source_session_timeout.py

```diff
@@ -1,17 +1,176 @@
-from . import functional_test
+import socket
+import time
+from dataclasses import dataclass
+from multiprocessing import Process
+from pathlib import Path
 
+from selenium.webdriver.firefox.webdriver import WebDriver
 
-class TestSourceSessions(
-        functional_test.FunctionalTest,
-        source_navigation_steps.SourceNavigationStepsMixin):
+import pytest
+import requests
 
-    session_expiration = 5
+from models import Journalist
+from tests.functional.app_navigators import SourceAppNagivator, WebDriverTypeEnum, get_web_driver
+from tests.functional.db_session import get_database_session
+from tests.functional.factories import SecureDropConfigFactory
+from tests.functional.sd_config_v2 import SecureDropConfig
 
-    def test_source_session_timeout(self):
-        self._source_visits_source_homepage()
-        self._source_clicks_submit_documents_on_homepage()
-        self._source_continues_to_submit_page()
-        self._source_waits_for_session_to_timeout()
-        self.driver.refresh()
-        self._source_sees_session_timeout_message()
+
+def _get_unused_port() -> int:
+    s = socket.socket()
+    s.bind(("127.0.0.1", 0))
+    port = s.getsockname()[1]
+    s.close()
+    return port
+
+
+def _start_source_server(port: int, config: SecureDropConfig) -> None:
+    # This function will be called in a separate Process that runs the source app
+    # Modify the sdconfig module in the app's memory so that it mirrors the supplied config
+    # Do this BEFORE importing any other module of the application so the modified config is
+    # what eventually gets imported by the app's code
+    import sdconfig
+
+    sdconfig.config = config  # type: ignore
+
+    # Then start the source app
+    from source_app import create_app
+
+    source_app = create_app(config)  # type: ignore
+    source_app.run(port=port, debug=True, use_reloader=False, threaded=True)
+
+
+def _start_journalist_server(port: int, config: SecureDropConfig) -> None:
+    # This function will be called in a separate Process that runs the journalist app
+    # Modify the sdconfig module in the app's memory so that it mirrors the supplied config
+    # Do this BEFORE importing any other module of the application so the modified config is
+    # what eventually gets imported by the app's code
+    import sdconfig
+
+    sdconfig.config = config  # type: ignore
+
+    # Then start the journalist app
+    from journalist_app import create_app
+
+    journalist_app = create_app(config)  # type: ignore
+    journalist_app.run(port=port, debug=True, use_reloader=False, threaded=True)
+
+
+@dataclass(frozen=True)
+class SdServersFixtureResult:
+    source_app_base_url: str
+    journalist_app_base_url: str
+
+
+# Very short session expiration time
+SESSION_EXPIRATION_SECONDS = 3
+
+
+# TOOD(AD): This could eventually replace the sd_servers fixture
+@pytest.fixture(scope="session")
+def sd_servers_with_short_session_timeout(setup_journalist_key_and_gpg_folder):
```

Comment:
How would you pass in a different SESSION_EXPIRATION_SECONDS in this fixture for tests that don't need a a short session?

Reply:
In the next PRs, I've made this code more generic so that it can be used for tests that require a normal session timeout. Here it's just an initial example.